helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -220,6 +271,274 @@ private MessageOutput throttlePerReplicaMessages(IdealState idealState,
+              partition.getPartitionName(), stateModelDef, cache);
+      currentStateCountsByPartition.put(partition, currentStateCounts);
+    }
+
+    return currentStateCountsByPartition;
+  }
+
+  /*
+   * Classify the messages of each partition into recovery and load messages.
+   */
+  private void classifyMessages(String resourceName, StateModelDefinition stateModelDef,
+      ResourceControllerDataProvider cache, Map<Partition, List<Message>> selectedResourceMessages,
+      List<Message> recoveryMessages, List<Message> loadMessages,
+      Map<Partition, Map<String, Integer>> expectedStateCountByPartition,
+      Map<Partition, Map<String, Integer>> currentStateCountsByPartition) {
+    logger.info("Event info {}, Classify message for resource {} ", _eventId, resourceName);
+
+    for (Partition partition : selectedResourceMessages.keySet()) {
+      Map<String, Integer> partitionExpectedStateCounts =
+          expectedStateCountByPartition.get(partition);
+      Map<String, Integer> partitionCurrentStateCounts =
+          currentStateCountsByPartition.get(partition);
+
+      List<Message> partitionMessages = selectedResourceMessages.get(partition);
+      if (partitionMessages == null) {
+        continue;
+      }
+
+      // sort partitionMessages based on transition priority and then creation timestamp for transition message
+      // TODO: sort messages in same partition in next PR
+      Set<String> disabledInstances =
+          cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName());
+      for (Message msg : partitionMessages) {
+        if (!Message.MessageType.STATE_TRANSITION.name().equals(msg.getMsgType())) {
+          logger.debug(
+              "Event info: {} Message: {} not subject to throttle in resource: {} with type {}",
+              _eventId, msg, resourceName, msg.getMsgType());
+          continue;
+        }
+
+        boolean isUpward = !isDownwardTransition(stateModelDef, msg);
+
+        // TODO: add disabled disabled instance special treatment
+
+        String toState = msg.getToState();
+
+        // TODO: dropped and Error state special treatment
+
+        Integer minimumRequiredCount = partitionExpectedStateCounts.getOrDefault(toState, 0);
+        Integer currentCount = partitionCurrentStateCounts.getOrDefault(toState, 0);
+
+        //
+        if (isUpward && (currentCount < minimumRequiredCount)) {
+          recoveryMessages.add(msg);
+          // It is critical to increase toState value by one here. For example, current state
+          // of 3 replica in a partition is (M, O, O). Two messages here bringing up the two O to S.
+          // In this case, the first O->S would be classified as recovery. Then this line would
+          // increase the S state in partitionCurrentStateCounts value by one. Next O->S message
+          // would be correctly marked as load message.
+          partitionCurrentStateCounts.put(toState, currentCount + 1);
```

Comment:
Do we need to propagate the count here too? Since the original input count was calculated with propagation, right?

Reply:
Yes, I think so. 

We can reason it this way:
The propagated current state counts only reflected what current state is. This line is more like saying, if this new message delivered and the new state change happens, what it will be. 

Let me what is your take here?

Reply:
I mean simply add one to the immediate toState is not propagation, right? Unless I take something wrong, but I think you should add all the further to states down the line here.